File: c/14/fourteen.c
#include <iostream>
#include <ostream>


bool even(int i)
{
  return i % 2 == 0;
}

int functor(int i, int call)
{
  if (i == 1)
  { 
    return 1+call;  
  }
  else if (even(i))
  {
    return functor( i/2, 1 + call);
  }
  else
  {
    return functor( 3*i+1, 1 + call);
  }
}

int main() 
{
  int largest = 0;
  for(int i = 999999; i >= 0; i--)
  {
    int cur = functor(i, 0);
    if (cur > largest)
    {
      largest = cur;
      std::cout << cur << std::endl;
    }
  }
}

File: clojure/14/fourteen.clj
(use 'clojure.contrib.map-utils)

(defn functor [i orig call]
  (if (== i 1)
    (hash-map (+ 1 call) orig )
    (if (even? i) 
      (recur (/ i 2) orig (+ 1 call))
      (recur (+ (* 3 i) 1) orig (+ 1 call)))))

(last (sort (apply merge (pmap (fn [x] (functor x x 0)) (range  799999 999999)))))
